src/osdc/Objecter.cc

```diff
@@ -1944,7 +1944,7 @@ void Objecter::_maybe_request_map()
     flag = CEPH_SUBSCRIBE_ONETIME;
   }
   epoch_t epoch = osdmap->get_epoch() ? osdmap->get_epoch()+1 : 0;
-  if (monc->sub_want("osdmap", epoch, flag)) {
+  if (monc->sub_want("osdmap", epoch, flag | CEPH_SUBSCRIBE_LATEST_OSDMAP)) {
```

Comment:
As mentioned previously, it's not just clients that use Objecter:

> Objecter is also used by managers, MDSes and even OSDs in some cases, I think. If the intention is to turn on CEPH_SUBSCRIBE_LATEST_OSDMAP behaviour just for clients, this is probably wrong.

Unless I'm missing something, this turns on the proposed behaviour for all those users.

Reply:
@idryomov  This problem is caused by the kernel-client. When I look at the code, I think MDS, Mgr are all clients of rados. So I modified objector instead of mdsclient(add flag to show whether this is a client or MDS )

Reply:
Well, this PR, if merged as is, would affect everything, not just the kernel client.

Reply:
The kernel client and ceph-fuse should behave the same when it comes to subscribing to osdmaps.  IIRC both the kernel client and ceph-fuse don't subscribe to continuous updates and therefore don't receive osdmap updates when the mount is idle.

If there is a discrepancy in behaviour there, and you don't see high CPU usage on the monitor with ceph-fuse, let's look into and fix the kernel client.

Reply:
@idryomov.  yes. I think fuse-client has the same problem.
Do you suggest that I only submit the first commit and kernel patch?

Reply:
> Blacklist events are consumed by the clients as well, see `Client::handle_osd_map()`. And again, blacklist handling is probably not the only place where we expect to see all osdmaps, so I don't want to go into that particular case, looking for workarounds. I don't think we are prepared to cope with missing osdmaps on a regular basis.
> 
> Tagging as core, as this change affects everything, not just the monitor. @jdurgin, @gregsfortytwo, any thoughts?

Looking through the mgr modules, none of them are affected by skipping osdmaps. That's the biggest risk of this change though - as Ilya points out it's not a common path today, so there may be lurking bugs that are difficult to detect. If we could address this client side that would be less risky.

Reply:
@jdurgin  
I will check mgr module. thanks.
There are two main reasons why we need to modify mon
1. The client doesn‘t know how much versions its osdmap is behind the osdmap of mon.
2. The way mon handles requests is to send to the client all versions of osdmap that are missing.

The solution of this pr is to add a flag(CEPH_SUBSCRIBE_LATEST) to indicate whether the client can handle a full map and skip the intermediate versions.
If the client's request doesn’t carry the flag, mon will follow the original processing method
If the client has this flag, it means that the client can accept a full osdmap, and mon will measure the cost of the full osdmap or the incremental osdmap according to the version gap of the client.
The modification on mon is mainly to recognize this flag.

Another scenario is to add a message to get the latest version of the current version. In this way, cleint can decide whether to request for a full map. However, this solution needs to give priority to upgrading monitor, and it needs to carefully design the time to ask for the latest version.

I would like to say that for all SSD clusters with thousands of nodes, the map version is growing rapidly.
The distribution of maps, I think, will be a factor limiting the size and stability of clusters.
Maybe adding a module to share the responsibility of the server to distribute osdmap is also a method (like mgr)

Reply:
@jdurgin 
I did not find any potential problems with the mgr module. The monclient and objecter are mainly called by its submodules to send commands to the mon and the cluster.
thx